java/api/src/main/java/com/exadel/frs/core/trainservice/service/ModelStatisticService.java

```diff
@@ -0,0 +1,122 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.transaction.Transactional;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.scheduling.support.CronExpression;
+import org.springframework.stereotype.Service;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class ModelStatisticService {
+
+    private static final String CRON_EXPRESSION = "0 0 */1 * * ?";
+
+    private final ModelRepository modelRepository;
+    private final TableLockRepository lockRepository;
+    private final ModelStatisticRepository statisticRepository;
+    private final ModelStatisticCacheProvider statisticCacheProvider;
+
+    @Transactional
+    @Scheduled(cron = CRON_EXPRESSION)
+    public void updateAndRecordStatistics() {
+        val cronStep = new CronStep();
+        // Used to obtain a table lock. Only one application instance per time can execute the method.
+        lockRepository.lockByName(MODEL_STATISTIC_LOCK);
+
+        val cache = statisticCacheProvider.getCacheCopyAsMap();
+        statisticCacheProvider.invalidateCache();
+
+        if (cache.isEmpty()) {
+            log.info("No statistic to update/record");
+            return;
+        }
+
+        val updatedStatistics = updateStatistics(cache, cronStep);
+        val recordedStatistics = recordStatistics(cache);
+
+        val updateCount = updatedStatistics.size();
+        val recordCount = recordedStatistics.size();
+
+        val statistics = new ArrayList<ModelStatistic>(updateCount + recordCount);
+        statistics.addAll(updatedStatistics);
+        statistics.addAll(recordedStatistics);
+
+        statisticRepository.saveAll(statistics);
+        log.info("The statistics have been recorded({}) and updated({})", recordCount, updateCount);
+    }
+
+    private List<ModelStatistic> updateStatistics(final Map<Long, ModelStatisticCacheEntry> cache,
+                                                  final CronStep cronStep) {
+        val updatedStatistics = new ArrayList<ModelStatistic>();
+
+        findStatisticsToUpdate(cache.keySet(), cronStep).forEach(statistic -> {
+            val cacheKey = statistic.getModel().getId();
+            val cacheEntry = cache.get(cacheKey);
+
+            val totalRequestCount = statistic.getRequestCount() + cacheEntry.getRequestCount();
+            statistic.setRequestCount(totalRequestCount);
+
+            updatedStatistics.add(statistic);
+            cache.remove(cacheKey);
+        });
+
+        return updatedStatistics;
+    }
+
+    private List<ModelStatistic> findStatisticsToUpdate(final Set<Long> modelIds,
+                                                        final CronStep cronStep) {
+        return statisticRepository.findAllByModelIdsAndCreatedDateBetween(
+                modelIds, cronStep.getCurrent(), cronStep.getNext());
```

Comment:
If we make the cron expression as an environment variable, we still want to see statistics each hour.
So if we sum up all my comments:
1. Truncate date to hour
2. Make cron a variable
It's enough to truncate cronStep time to an hour and use it for search.